### PR TITLE
fix: Limit the length of Event#parameters(#10263)

### LINKED ISSUE #10263: [Bug] Event List information Query data exception
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

OAP server (apache/skywalking)

### What happened

The agent collects data and the server cannot serialize it with json

### What you expected to happen

event list Normal query

### How to reproduce

Using JVM information:
IBM J9 VM (build 2.9, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20211022_15212 (JIT enabled, AOT enabled) IBM J9 VM (Build 2.9, JRE 1.8.0 Linux AMD64-64-bit Compressed References 20211022_15212 (JIT Enabled, aOT Enabled)

```
2023-01-11 09:02:30,628 notprivacysafe.graphql.execution.SimpleDataFetcherExceptionHandler 27 [armeria-eventloop-epoll-4-52] WARN [] - Exception while fetching data (/data) : com.google.gson.stream.MalformedJsonException: Unterminated string at line 1 column 2001 path $.
com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated string at line 1 column 2001 path $.
at com.google.gson.Gson.fromJson(Gson.java:1006) ~[gson-2.9.0.jar:?]
at com.google.gson.Gson.fromJson(Gson.java:956) ~[gson-2.9.0.jar:?]
at com.google.gson.Gson.fromJson(Gson.java:905) ~[gson-2.9.0.jar:?]
at org.apache.skywalking.oap.server.core.query.type.event.Event.setParameters(Event.java:58) ~[server-core-9.2.0.jar:9.2.0]
at org.apache.skywalking.oap.server.storage.plugin.elasticsearch.query.ESEventQueryDAO.parseSearchHit(ESEventQueryDAO.java:179) ~[storage-elasticsearch-plugin-9.2.0.jar:9.2.0]
at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source) ~[?:?]
at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
at org.apache.skywalking.oap.server.storage.plugin.elasticsearch.query.ESEventQueryDAO.getEventsResultByCurrentBuilder(ESEventQueryDAO.java:75) ~[storage-elasticsearch-plugin-9.2.0.jar:9.2.0]
at org.apache.skywalking.oap.server.storage.plugin.elasticsearch.query.ESEventQueryDAO.queryEvents(ESEventQueryDAO.java:58) ~[storage-elasticsearch-plugin-9.2.0.jar:9.2.0]
at org.apache.skywalking.oap.server.core.query.EventQueryService.queryEvents(EventQueryService.java:56) ~[server-core-9.2.0.jar:9.2.0]
at org.apache.skywalking.oap.query.graphql.resolver.EventQuery.queryEvents(EventQuery.java:49) ~[query-graphql-plugin-9.2.0.jar:9.2.0]
at jdk.internal.reflect.GeneratedMethodAccessor167.invoke(Unknown Source) ~[?:?]
at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
at java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
at graphql.kickstart.tools.resolver.MethodFieldResolverDataFetcher.get(MethodFieldResolver.kt:280) ~[graphql-java-tools-12.0.2.jar:?]
at graphql.execution.ExecutionStrategy.fetchField(ExecutionStrategy.java:279) ~[graphql-java-17.3.jar:?]
at graphql.execution.ExecutionStrategy.resolveFieldWithInfo(ExecutionStrategy.java:210) ~[graphql-java-17.3.jar:?]
at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:60) ~[graphql-java-17.3.jar:?]
at graphql.execution.Execution.executeOperation(Execution.java:159) ~[graphql-java-17.3.jar:?]
at graphql.execution.Execution.execute(Execution.java:105) ~[graphql-java-17.3.jar:?]
at graphql.GraphQL.execute(GraphQL.java:613) ~[graphql-java-17.3.jar:?]
at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:538) ~[graphql-java-17.3.jar:?]
at graphql.GraphQL.executeAsync(GraphQL.java:502) ~[graphql-java-17.3.jar:?]
at com.linecorp.armeria.server.graphql.DefaultGraphqlService.execute(DefaultGraphqlService.java:101) ~[armeria-graphql-1.16.0.jar:?]
at com.linecorp.armeria.server.graphql.DefaultGraphqlService.executeGraphql(DefaultGraphqlService.java:93) ~[armeria-graphql-1.16.0.jar:?]
at com.linecorp.armeria.server.graphql.protocol.AbstractGraphqlService.lambda$doPost$0(AbstractGraphqlService.java:117) ~[armeria-graphql-protocol-1.16.0.jar:?]
at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?]
at com.linecorp.armeria.common.util.UnmodifiableFuture.doComplete(UnmodifiableFuture.java:152) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.util.UnmodifiableFuture.lambda$wrap$0(UnmodifiableFuture.java:103) ~[armeria-1.16.0.jar:?]
at java.util.concurrent.CompletableFuture.uniHandle(Unknown Source) ~[?:?]
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?]
at com.linecorp.armeria.common.stream.StreamMessageCollector.onComplete(StreamMessageCollector.java:67) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:258) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent0(DefaultStreamMessage.java:300) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:292) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:429) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:372) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.doRequest(DefaultStreamMessage.java:275) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.request(DefaultStreamMessage.java:257) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.AbstractStreamMessage$SubscriptionImpl.request(AbstractStreamMessage.java:211) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.StreamMessageCollector.onSubscribe(StreamMessageCollector.java:52) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.subscribe(DefaultStreamMessage.java:149) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.stream.DefaultStreamMessage.lambda$subscribe$0(DefaultStreamMessage.java:135) ~[armeria-1.16.0.jar:?]
at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:547) ~[armeria-1.16.0.jar:?]
at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174) [netty-common-4.1.77.Final.jar:4.1.77.Final]
at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167) [netty-common-4.1.77.Final.jar:4.1.77.Final]
at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) [netty-common-4.1.77.Final.jar:4.1.77.Final]
at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:391) [netty-transport-classes-epoll-4.1.77.Final.jar:4.1.77.Final]
at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995) [netty-common-4.1.77.Final.jar:4.1.77.Final]
at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.77.Final.jar:4.1.77.Final]
at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.77.Final.jar:4.1.77.Final]
at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: com.google.gson.stream.MalformedJsonException: Unterminated string at line 1 column 2001 path $.
at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1597) ~[gson-2.9.0.jar:?]
at com.google.gson.stream.JsonReader.nextQuotedValue(JsonReader.java:1031) ~[gson-2.9.0.jar:?]
at com.google.gson.stream.JsonReader.nextString(JsonReader.java:815) ~[gson-2.9.0.jar:?]
at com.google.gson.internal.bind.TypeAdapters$15.read(TypeAdapters.java:393) ~[gson-2.9.0.jar:?]
at com.google.gson.internal.bind.TypeAdapters$15.read(TypeAdapters.java:381) ~[gson-2.9.0.jar:?]
at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41) ~[gson-2.9.0.jar:?]
at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187) ~[gson-2.9.0.jar:?]
at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145) ~[gson-2.9.0.jar:?]
at com.google.gson.Gson.fromJson(Gson.java:991) ~[gson-2.9.0.jar:?]
... 60 more
```

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by leon756**
### Fix <bug description or the bug issue number or bug issue link>
- [x] Add a unit test to verify that the fix works.
- [x] Explain briefly why the bug exists and how to fix it:
        Keep the JSON format when substring
  

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #10263.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

File: oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/StringUtilTest.java [hunk 1]
```diff
@@ -76,4 +76,12 @@ public void testTrim() {
         assertEquals(StringUtil.trim("abcdef", 'a'), "bcdef");
         assertEquals(StringUtil.trim("abcdef", 'f'), "abcde");
     }
+
+    @Test
+    public void testTrimJson() {
+        String jsonString = "{\"k1\":\"v1\",\"k2\":\"v2\",\"k3\":true,\"k4\":[{\"a\":\"b\",\"c\":\"d\"},{\"e\":\"f\",\"g\":\"h\"}]}";
+        assertEquals(StringUtil.trimJson(jsonString, 10), "{\"k1\":\"v1\",\"k2\":\"v2\"}");
+        assertEquals(StringUtil.trimJson(jsonString, 30), "{\"k1\":\"v1\",\"k2\":\"v2\",\"k3\":true}");
+    }
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/StringUtilTest.java:85]: I think we should verify whether the size of the trimmed JSON str oversized.

Comment (leon756) [reply oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/StringUtilTest.java:92]: Sorry, I was careless.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/Event.java [hunk 2]
```diff
@@ -153,8 +154,15 @@ public boolean combine(final Metrics metrics) {
...
      */
     public void setParameters(String parameters) {
-        this.parameters = parameters == null || parameters.length() <= PARAMETER_MAX_LENGTH ?
-            parameters : parameters.substring(0, PARAMETER_MAX_LENGTH);
+        if (parameters == null || parameters.length() <= PARAMETER_MAX_LENGTH) {
+            this.parameters = parameters;
+        } else {
+            try {
+                this.parameters = StringUtil.trimJson(parameters, PARAMETER_MAX_LENGTH);
+            } catch (Exception e) {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/Event.java:162]: I can't see the point of this catch. What do you want to do about this case? When is the exception thrown?

Comment (leon756) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/Event.java:162]: I am not quite sure that the original parameters must be in json format

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/Event.java:162]: @kezhenxu94 Is the JSON format required by the protocol?

Comment (leon756) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/Event.java:162]: If not, `JsonParser.parseString(jsonString)` may throw `JsonSyntaxException`

File: oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/StringUtilTest.java [hunk 3]
```diff
@@ -76,4 +76,19 @@ public void testTrim() {
         assertEquals(StringUtil.trim("abcdef", 'a'), "bcdef");
         assertEquals(StringUtil.trim("abcdef", 'f'), "abcde");
     }
+
+    @Test
+    public void testTrimJson() {
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/StringUtilTest.java:81]: This UT is testing flat JSON, how about JSON with a hierarchy structure?

**Comment by wu-sheng** [comment]
@kezhenxu94 I also want to push a little back about this fix. What do you think we make this longer, than in the storage level, it would be a `text` rather than a varchar. Considering `storageOnly = true`, it should be harmless. Then we don't need this fix, but a type change only.

**Comment by kezhenxu94** [comment]
> @kezhenxu94 I also want to push a little back about this fix. What do you think we make this longer, than in the storage level, it would be a `text` rather than a varchar. Considering `storageOnly = true`, it should be harmless. Then we don't need this fix, but a type change only.

Changing the type of Event#parameters is good for ElasticSearch storage, but for JDBC storage it seems to have no effect.

**Comment by wu-sheng** [comment]
> > @kezhenxu94 I also want to push a little back about this fix. What do you think we make this longer, than in the storage level, it would be a `text` rather than a varchar. Considering `storageOnly = true`, it should be harmless. Then we don't need this fix, but a type change only.
> 
> Changing the type of Event#parameters is good for ElasticSearch storage, but for JDBC storage it seems to have no effect.

I remember if it is oversize, it should be `text` type in JDBC. Isn't it?

**Comment by kezhenxu94** [comment]
> > > @kezhenxu94 I also want to push a little back about this fix. What do you think we make this longer, than in the storage level, it would be a `text` rather than a varchar. Considering `storageOnly = true`, it should be harmless. Then we don't need this fix, but a type change only.
> > 
> > 
> > Changing the type of Event#parameters is good for ElasticSearch storage, but for JDBC storage it seems to have no effect.
> 
> I remember if it is oversize, it should be `text` type in JDBC. Isn't it?

Only field of type `byte[]` is `mediumtext` in JDBC storage.

https://github.com/apache/skywalking/blob/ffc8453332bc991e97ad4c745e56437b5464e875/oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/common/JDBCTableInstaller.java#L107-L129

**Comment by wu-sheng** [comment]
This is for JDBC(H2), check MySQLTableInstaller

**Comment by wu-sheng** [comment]
@leon756 Are you following this?

**Comment by wu-sheng** [comment]
Fixed by https://github.com/apache/skywalking/pull/11032

**Pull request closed without merging**